{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "de43a647-845b-481d-94e3-4c643439b96a",
   "metadata": {},
   "source": [
    "## Create DB and analyze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c60fb8d8-34df-4d54-9149-b27144726ef4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import csv\n",
    "import math\n",
    "from Bio import AlignIO\n",
    "from collections import Counter\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ce215f8a-27ce-466c-8d12-7afd009ca97e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONNECT to the database\n",
    "conn = sqlite3.connect('compar_gen_data.db')\n",
    "cursor = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "11184a74-adca-4bf6-a61a-e341d6ed230d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all functions\n",
    "\n",
    "# Create tables\n",
    "def create_tables():\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    # Create Gene table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS Gene (\n",
    "            gene_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            gene_name TEXT UNIQUE,\n",
    "            target_seq TEXT;\n",
    "        )\n",
    "    ''')\n",
    "\n",
    "    # Create Species table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS Species (\n",
    "            species_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            species_name TEXT UNIQUE\n",
    "        )\n",
    "    ''')\n",
    "    \n",
    "    # Create Mutation table\n",
    "            # Mutation.edit_distance is a number of nucleotide substitutions needed in codon to move from ancestral amino acid to variant aa\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS Mutation (\n",
    "            mutation_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            gene_id INTEGER,\n",
    "            species_id INTEGER,\n",
    "            position INTEGER,\n",
    "            ancestral_residue TEXT,\n",
    "            variant_residue TEXT,\n",
    "            edit_distance INTEGER,\n",
    "            FOREIGN KEY (gene_id) REFERENCES Gene(gene_id),\n",
    "            FOREIGN KEY (species_id) REFERENCES Species(species_id)\n",
    "        )\n",
    "    ''')\n",
    "\n",
    "    # Create DMS table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS DMS (\n",
    "            dms_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            mutation_id INTEGER,\n",
    "            score REAL,\n",
    "            FOREIGN KEY (mutation_id) REFERENCES Mutation(mutation_id)\n",
    "        )\n",
    "    ''')\n",
    "\n",
    "    # Create MSA table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS MSA (\n",
    "            msa_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            mutation_id INTEGER,\n",
    "            shannon_entropy REAL,\n",
    "            jsd REAL,\n",
    "            phylop REAL,\n",
    "            phastcons REAL,\n",
    "            gerp REAL,\n",
    "            percentage_identity REAL,\n",
    "            ci REAL,\n",
    "            variant_percentage_residue REAL,\n",
    "            FOREIGN KEY (mutation_id) REFERENCES Mutation(mutation_id)\n",
    "        )\n",
    "    ''')\n",
    "\n",
    "    # Create IntegratedData table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS IntegratedData (\n",
    "            integrated_data_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            mutation_id INTEGER,\n",
    "            dms_score REAL,\n",
    "            shannon_entropy REAL,\n",
    "            jsd REAL,\n",
    "            phylop REAL,\n",
    "            phastcons REAL,\n",
    "            gerp REAL,\n",
    "            percentage_identity REAL,\n",
    "            ci REAL,\n",
    "            variant_percentage_residue REAL,\n",
    "            FOREIGN KEY (mutation_id) REFERENCES Mutation(mutation_id),\n",
    "            FOREIGN KEY (dms_score) REFERENCES DMS(score),\n",
    "            FOREIGN KEY (shannon_entropy) REFERENCES MSA(shannon_entropy),\n",
    "            FOREIGN KEY (jsd) REFERENCES MSA(jsd),\n",
    "            FOREIGN KEY (phylop) REFERENCES MSA(phylop),\n",
    "            FOREIGN KEY (phastcons) REFERENCES MSA(phastcons),\n",
    "            FOREIGN KEY (gerp) REFERENCES MSA(gerp),\n",
    "            FOREIGN KEY (percentage_identity) REFERENCES MSA(percentage_identity),\n",
    "            FOREIGN KEY (ci) REFERENCES MSA(ci),\n",
    "            FOREIGN KEY (variant_percentage_residue) REFERENCES MSA(variant_percentage_residue)\n",
    "        )\n",
    "    ''')\n",
    "\n",
    "    # Create SummaryStatistics table\n",
    "    cursor.execute('''\n",
    "        CREATE TABLE IF NOT EXISTS SummaryStatistics (\n",
    "            summary_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "            gene_id INTEGER,\n",
    "            mean_conservation_score REAL,\n",
    "            mean_discordance REAL,\n",
    "            shannon_entropy REAL,\n",
    "            FOREIGN KEY (gene_id) REFERENCES Gene(gene_id)\n",
    "        )\n",
    "    ''')\n",
    "    conn.commit()\n",
    "\n",
    "# Insert genes into the Gene table\n",
    "def insert_genes(gene):\n",
    "    for g in gene:\n",
    "        cursor.execute('''\n",
    "            INSERT OR IGNORE INTO Gene (gene_name)\n",
    "            VALUES (?)\n",
    "        ''', (g,))\n",
    "    conn.commit()\n",
    "\n",
    "# Parse Clustal W file and extract mutation data\n",
    "def parse_clustalw(gene, file_path):\n",
    "    alignment = AlignIO.read(file_path, \"clustal\")\n",
    "    alignment_data = []\n",
    "    for record in alignment:\n",
    "        gene_species, sequence = record.id.split('/', 1)\n",
    "        species_name = gene_species\n",
    "        # Check if gene name already exists in Genes table\n",
    "        for gene_name in [gene]: \n",
    "            cursor.execute('SELECT gene_id FROM Gene WHERE gene_name = ?', (gene_name,))\n",
    "            result = cursor.fetchone()\n",
    "            if result:\n",
    "                gene_id = result[0]\n",
    "            else:\n",
    "                # Insert gene name into Genes table if it doesn't exist\n",
    "                cursor.execute('INSERT INTO Gene (gene_name) VALUES (?)', (gene_name,))\n",
    "                gene_id = cursor.lastrowid\n",
    "        for position, (ancestral_residue, residue) in enumerate(zip(alignment[0].seq, record.seq), start=1):\n",
    "            alignment_data.append((gene_id, species_name, position, ancestral_residue, residue))\n",
    "    return alignment_data\n",
    "\n",
    "# Insert mutation data into the Mutation table\n",
    "def insert_alignment_data(alignment_data):\n",
    "    for gene_id, species_name, position, ancestral_residue, residue in alignment_data:\n",
    "        # Check if species name already exists in Species table\n",
    "        cursor.execute('SELECT species_id FROM Species WHERE species_name = ?', (species_name,))\n",
    "        result = cursor.fetchone()\n",
    "        if result:\n",
    "            species_id = result[0]\n",
    "        else:\n",
    "            # Insert species name into Species table if it doesn't exist\n",
    "            cursor.execute('INSERT INTO Species (species_name) VALUES (?)', (species_name,))\n",
    "            species_id = cursor.lastrowid\n",
    "        # Insert mutation data into Mutation table\n",
    "        cursor.execute('''\n",
    "            INSERT INTO Mutation (gene_id, species_id, position, ancestral_residue, variant_residue)\n",
    "            VALUES (?, ?, ?, ?, ?)\n",
    "        ''', (gene_id, species_id, position, ancestral_residue, residue))\n",
    "        # Get the last inserted mutation_id\n",
    "        mutation_id = cursor.lastrowid\n",
    "        \n",
    "        # Insert mutation data into MSA table\n",
    "        cursor.execute('''\n",
    "            INSERT INTO MSA (mutation_id, shannon_entropy, jsd, phylop, phastcons, gerp, percentage_identity, ci)\n",
    "            VALUES (?, ?, ?, ?, ?, ?, ?, ?)\n",
    "        ''', (mutation_id, None, None, None, None, None, None, None))  # Replace None with your data\n",
    "    conn.commit()\n",
    "\n",
    "# Function to insert DMS data into the Mutation and DMS tables\n",
    "def insert_dms_data(gene, species, file_path, col_position, col_ancestral, col_variant, col_score):\n",
    "    # Open CSV file containing DMS data\n",
    "    with open(file_path, 'r', newline='') as csvfile:\n",
    "        reader = csv.DictReader(csvfile)\n",
    "        dms_data = []\n",
    "        for row in reader:\n",
    "            position = int(row[col_position])\n",
    "            ancestral_residue = row[col_ancestral]\n",
    "            variant_residue = row[col_variant]\n",
    "            try:\n",
    "                score = float(row[col_score])\n",
    "            except ValueError:\n",
    "                #print(f\"Skipping row with invalid score: {row}\")\n",
    "                continue  # skip invalid scores\n",
    "\n",
    "            # Check if the species already exists in the Species table\n",
    "            cursor.execute('SELECT species_id FROM Species WHERE species_name = ?', (species,))\n",
    "            species_result = cursor.fetchone()\n",
    "            if species_result:\n",
    "                species_id = species_result[0]\n",
    "            else:\n",
    "                # Insert the species into the Species table if it doesn't exist\n",
    "                cursor.execute('INSERT INTO Species (species_name) VALUES (?)', (species,))\n",
    "                species_id = cursor.lastrowid\n",
    "                \n",
    "            # Check if the mutation already exists in the Mutation table\n",
    "            cursor.execute('''\n",
    "                SELECT mutation_id FROM Mutation \n",
    "                WHERE position = ? AND ancestral_residue = ? AND variant_residue = ?\n",
    "            ''', (position, ancestral_residue, variant_residue))\n",
    "            result = cursor.fetchone()\n",
    "            if result:\n",
    "                mutation_id = result[0]\n",
    "            else:\n",
    "                # Insert the mutation into the Mutation table if it doesn't exist\n",
    "                cursor.execute('''\n",
    "                    INSERT INTO Mutation (gene_id, species_id, position, ancestral_residue, variant_residue)\n",
    "                    VALUES (\n",
    "                        (SELECT gene_id FROM Gene WHERE gene_name = ?),\n",
    "                        (SELECT species_id FROM Species WHERE species_name = ?),\n",
    "                        ?, ?, ?\n",
    "                    )\n",
    "                ''', (gene, species, position, ancestral_residue, variant_residue))\n",
    "                mutation_id = cursor.lastrowid\n",
    "            # Append DMS data\n",
    "            dms_data.append((mutation_id, score))\n",
    "    \n",
    "    # Insert DMS data into the DMS table\n",
    "    cursor.executemany('''\n",
    "        INSERT INTO DMS (mutation_id, score)\n",
    "        VALUES (?, ?)\n",
    "    ''', dms_data)\n",
    "    conn.commit()\n",
    "\n",
    "# Calculate the Shannon entropy of a sequence\n",
    "def calculate_shannon_entropy(sequence):\n",
    "    counts = Counter(sequence)\n",
    "    total = len(sequence)\n",
    "    entropy = -sum((count / total) * math.log2(count / total) for count in counts.values())\n",
    "    return entropy\n",
    "\n",
    "# Calculate Percentage Identity\n",
    "def calculate_percentage_identity(residues, total_sequences):\n",
    "    most_common_residue_count = Counter(residues).most_common(1)[0][1]\n",
    "    percentage_identity = most_common_residue_count / total_sequences\n",
    "    return percentage_identity\n",
    "\n",
    "# Calculate variant_percentage_residue - % of each variant amino acid within the same position in MSA\n",
    "def calculate_variant_percentage_residue(gene_id, position, variant_residue):\n",
    "    cursor.execute('''\n",
    "        SELECT COUNT(*)\n",
    "        FROM Mutation\n",
    "        WHERE gene_id = ? AND position = ? AND variant_residue = ?\n",
    "    ''', (gene_id, position, variant_residue))\n",
    "    variant_residue_count = cursor.fetchone()[0]\n",
    "    \n",
    "    cursor.execute('''\n",
    "        SELECT COUNT(DISTINCT species_id)\n",
    "        FROM Mutation\n",
    "        WHERE gene_id = ? AND position = ?\n",
    "    ''', (gene_id, position))\n",
    "    total_unique_species_count = cursor.fetchone()[0]\n",
    "    \n",
    "    if total_unique_species_count == 0:\n",
    "        return 0\n",
    "    return variant_residue_count / total_unique_species_count\n",
    "\n",
    "# Fill variant_percentage_residue in MSA table\n",
    "def fill_variant_percentage_residue_in_msa():\n",
    "    # Fetch all mutation data required for calculation\n",
    "    cursor.execute('''\n",
    "        SELECT Mutation.mutation_id, Mutation.gene_id, Mutation.position, Mutation.variant_residue\n",
    "        FROM Mutation\n",
    "        JOIN MSA ON Mutation.mutation_id = MSA.mutation_id\n",
    "    ''')\n",
    "    mutation_data = cursor.fetchall()\n",
    "\n",
    "    # Calculate and update variant_percentage_residue for each mutation\n",
    "    for mutation_id, gene_id, position, variant_residue in mutation_data:\n",
    "        variant_percentage_residue = calculate_variant_percentage_residue(gene_id, position, variant_residue)\n",
    "        \n",
    "        cursor.execute('''\n",
    "            UPDATE MSA\n",
    "            SET variant_percentage_residue = ?\n",
    "            WHERE mutation_id = ?\n",
    "        ''', (variant_percentage_residue, mutation_id))\n",
    "    \n",
    "    conn.commit()\n",
    "\n",
    "def update_consevation_scores():\n",
    "    # Retrieve data from Mutation table\n",
    "    cursor.execute('''\n",
    "        SELECT Mutation.gene_id, Mutation.position, Mutation.ancestral_residue, Mutation.variant_residue, Species.species_name\n",
    "        FROM Mutation\n",
    "        JOIN MSA ON Mutation.mutation_id = MSA.mutation_id\n",
    "        JOIN Species ON Mutation.species_id = Species.species_id\n",
    "    ''')\n",
    "    mutation_data = cursor.fetchall()\n",
    "    \n",
    "    # Group data by gene_id and position to calculate Shannon entropy\n",
    "    gene_position_data = {}\n",
    "    unique_species_count = {}\n",
    "    for gene_id, position, ancestral_residue, variant_residue, species_name in mutation_data:\n",
    "        if (gene_id, position) not in gene_position_data:\n",
    "            gene_position_data[(gene_id, position)] = []\n",
    "            unique_species_count[(gene_id, position)] = set()  # Store unique species for each (gene_id, position)\n",
    "        gene_position_data[(gene_id, position)].append(variant_residue)\n",
    "        unique_species_count[(gene_id, position)].add(species_name)  # Add species to set\n",
    "        \n",
    "    # Calculate conservation scores for each gene_id and position group, and update the MSA table\n",
    "    for (gene_id, position), residues in gene_position_data.items():\n",
    "        entropy = calculate_shannon_entropy(residues)\n",
    "        total_sequences = len(unique_species_count[(gene_id, position)])\n",
    "        percentage_identity = calculate_percentage_identity(residues, total_sequences)        \n",
    "        \n",
    "        cursor.execute('''\n",
    "            UPDATE MSA\n",
    "            SET shannon_entropy = ?, percentage_identity = ?\n",
    "            WHERE mutation_id IN (\n",
    "                SELECT mutation_id\n",
    "                FROM Mutation\n",
    "                WHERE gene_id = ? AND position = ?\n",
    "            )\n",
    "        ''', (entropy, percentage_identity, gene_id, position))\n",
    "    conn.commit()\n",
    "\n",
    "def fill_integrated_data():\n",
    "    cursor.execute('''\n",
    "        INSERT INTO IntegratedData (mutation_id, dms_score, shannon_entropy, jsd, phylop, phastcons, gerp, percentage_identity, ci, variant_percentage_residue)\n",
    "        SELECT MSA.mutation_id, DMS.score, MSA.shannon_entropy, MSA.jsd, MSA.phylop, MSA.phastcons, MSA.gerp, MSA.percentage_identity, MSA.ci, MSA.variant_percentage_residue\n",
    "        FROM MSA\n",
    "        JOIN DMS ON MSA.mutation_id = DMS.mutation_id\n",
    "    ''')\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1be3207b-456b-454e-97b0-a56b6c4d77c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CREATE tables\n",
    "create_tables()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "b9cb942d-f74d-4788-94df-25aa450b826e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SET PTEN variables\n",
    "dms_species_name = 'Human'\n",
    "gene_name = 'PTEN'\n",
    "ancestral_sequence = (\n",
    "    \"MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVYRNNIDDVVRFLDSKHKNHYKIYNLCAERHYDTAKFNCRVAQYPFEDHNPPQLELIKPFCEDLDQWLSEDDNHVAAIHCKAGKGRTGVMICAYLLHRGKFLKAQEALDFYGEVRTRDKKGVTIPSQRRYVYYYSYLLKNHLDYRPVALLFHKMMFETIPMFSGGTC-----------------------NPQFVVCQLKVKIYSSNSGPTRREDKFMYFEFPQPLPVCGDIKVEFFHKQNKMLKKDKMFHFWVNTFFIPGPEETSEKVENGSLCDQEIDSICSIERADNDKEYLVLTLTKNDLDKANKDKANRYFSPNFKVKLYFTKTVEEPSNPEASSSTSVTPDVSDNEPDHYRYSDTTDSDPENEPFDEDQHTQITKV-\"\n",
    ")\n",
    "\n",
    "# MSA data\n",
    "aln_file_path = 'data/Human_PTEN_ortholog_prot_alignment_zebrafish.aln'\n",
    "\n",
    "# DMS data\n",
    "dms_file_path = 'data/dms-PTEN-013-a-1_1letter.csv'\n",
    "col_position = 'position'\n",
    "col_ancestral = 'ancestral_residue'\n",
    "col_variant = 'variant_residue'\n",
    "col_score = 'score'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3de636e5-e1ca-4110-9ab5-ce77a9b3327d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SET PAX6 variables\n",
    "dms_species_name = 'Human'\n",
    "gene_name = 'PAX6 Homeobox domain'\n",
    "ancestral_sequence = (\n",
    "    \"NRTSFTQEQIEALEKEFERTHYPDVFARERLAAKIDLPEARIQVWFSNRRAKWRR\"\n",
    ")\n",
    "\n",
    "# MSA data\n",
    "aln_file_path = 'data/Human_PAX6_Homeobox_domain_orthologues.aln'\n",
    "\n",
    "# DMS data\n",
    "dms_file_path = 'data/dms-PAX6-949-a-1-1letter.csv'\n",
    "col_position = 'position'\n",
    "col_ancestral = 'ancestral_residue'\n",
    "col_variant = 'variant_residue'\n",
    "col_score = 'score'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "68f5a30b-cafc-44e0-967f-0da52381d3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SET CYP2C19 variables\n",
    "dms_species_name = 'Human'\n",
    "gene_name = 'CYP2C19'\n",
    "ancestral_sequence = (\n",
    "    \"MDPFVVLVLCLSCLLLLSIWRQSSGRGKLPPGPTPLPVIGNILQIDIKDVSKSLTNLSKIYGPVFTLYFGLERMVVLHGYEVVKEALIDLGEEFSGRGHFPLAERANRGFGIVFSNGKRWKEIRRFSLMTLRNFGMGKRSIEDRVQEEARCLVEELRKTKASPCDPTFILGCAPCNVICSIIFQKRFDYKDQQFLNLMEKLNENIRIVSTPWIQICNNFPTIIDYFPGTHNKLLKNLAFMESDILEKVKEHQESMDINNPRDFIDCFLIKMEKEKQNQQSEFTIENLVITAADLLGAGTETTSTTLRYALLLLLKHPEVTAKVQEEIERVIGRNRSPCMQDRGHMPYTDAVVHEVQRYIDLIPTSLPHAVTCDVKFRNYLIPKGTTILTSLTSVLHDNKEFPNPEMFDPRHFLDEGGNFKKSNYFMPFSAGKRICVGEGLARMELFLFLTFILQNFNLKSLIDPKDLDTTPVVNGFASVPPFYQLCFIPV\"\n",
    ")\n",
    "\n",
    "# MSA data\n",
    "aln_file_path = 'data/Human_CYP2C19_orthologues.aln'\n",
    "\n",
    "# DMS data\n",
    "dms_file_path = 'data/dms-CYP2C19-1199-a-1.csv'\n",
    "col_position = 'position'\n",
    "col_ancestral = 'ancestral_residue'\n",
    "col_variant = 'variant_residue'\n",
    "col_score = 'score'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "534c178e-2215-4648-ae1b-c0da67c81803",
   "metadata": {},
   "outputs": [],
   "source": [
    "#FILL Mutation and MSA table from alignment file\n",
    "alignment_data = parse_clustalw(gene_name, aln_file_path)\n",
    "insert_alignment_data(alignment_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06621df6-887e-4b12-924b-2a4f8ae5195a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CALCULATE conservation scores and INSERT in MSA\n",
    "update_consevation_scores()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7dd8d7df-5c37-4dd9-ac2b-179aa6769a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FILL DMS data into the database\n",
    "insert_dms_data(gene_name, dms_species_name, dms_file_path, col_position, col_ancestral, col_variant, col_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "43f14ba1-be7f-4195-bcd7-06b45d28704c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FILL IntegratedData table\n",
    "fill_integrated_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7133bab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fill_variant_percentage_residue_in_msa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fe5c3957-f6c6-46aa-bc88-15ef09c8c6ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit_distance calculation\n",
    "codon_table = {\n",
    "    'F': ['TTT', 'TTC'],\n",
    "    'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "    'I': ['ATT', 'ATC', 'ATA'],\n",
    "    'M': ['ATG'],\n",
    "    'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "    'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "    'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "    'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "    'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "    'Y': ['TAT', 'TAC'],\n",
    "    'H': ['CAT', 'CAC'],\n",
    "    'Q': ['CAA', 'CAG'],\n",
    "    'N': ['AAT', 'AAC'],\n",
    "    'K': ['AAA', 'AAG'],\n",
    "    'D': ['GAT', 'GAC'],\n",
    "    'E': ['GAA', 'GAG'],\n",
    "    'C': ['TGT', 'TGC'],\n",
    "    'W': ['TGG'],\n",
    "    'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "    'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "    '*': ['TAA', 'TAG', 'TGA']\n",
    "}\n",
    "\n",
    "def one_nucleotide_away(codon1, codon2):\n",
    "    return sum(1 for a, b in zip(codon1, codon2) if a != b) == 1\n",
    "\n",
    "amino_acid_mapping = defaultdict(list)\n",
    "\n",
    "for aa1, codons1 in codon_table.items():\n",
    "    for codon1 in codons1:\n",
    "        for aa2, codons2 in codon_table.items():\n",
    "            if aa1 != aa2:\n",
    "                for codon2 in codons2:\n",
    "                    if one_nucleotide_away(codon1, codon2):\n",
    "                        amino_acid_mapping[aa1].append(aa2)\n",
    "                        break\n",
    "                        \n",
    "amino_acid_mapping = dict(amino_acid_mapping)\n",
    "\n",
    "# Fetch all mutations\n",
    "cursor.execute('SELECT mutation_id, ancestral_residue, variant_residue FROM Mutation')\n",
    "mutations = cursor.fetchall()\n",
    "\n",
    "# Update edit_distance for each mutation\n",
    "for mutation_id, ancestral_residue, variant_residue in mutations:\n",
    "    if variant_residue in amino_acid_mapping.get(ancestral_residue, []):\n",
    "        edit_distance = 1\n",
    "    else:\n",
    "        edit_distance = 'NA'\n",
    "    \n",
    "    cursor.execute('''\n",
    "        UPDATE Mutation\n",
    "        SET edit_distance = ?\n",
    "        WHERE mutation_id = ?\n",
    "    ''', (edit_distance, mutation_id))\n",
    "\n",
    "conn.commit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "da0e319a-5ad5-420a-9f4f-8387879ff16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temporary\n",
    "def update_variant_percentage_residue_in_integrated_data():\n",
    "    cursor.execute('''\n",
    "        UPDATE IntegratedData\n",
    "        SET variant_percentage_residue = (\n",
    "            SELECT MSA.variant_percentage_residue\n",
    "            FROM MSA\n",
    "            WHERE MSA.mutation_id = IntegratedData.mutation_id\n",
    "        )\n",
    "        WHERE EXISTS (\n",
    "            SELECT 1\n",
    "            FROM MSA\n",
    "            WHERE MSA.mutation_id = IntegratedData.mutation_id\n",
    "        )\n",
    "    ''')\n",
    "    conn.commit()\n",
    "\n",
    "# Run the function to update the IntegratedData table\n",
    "update_variant_percentage_residue_in_integrated_data()\n",
    "\n",
    "#cursor.execute('''    \n",
    "#    UPDATE Gene\n",
    "#    SET target_seq = \"MDPFVVLVLCLSCLLLLSIWRQSSGRGKLPPGPTPLPVIGNILQIDIKDVSKSLTNLSKIYGPVFTLYFGLERMVVLHGYEVVKEALIDLGEEFSGRGHFPLAERANRGFGIVFSNGKRWKEIRRFSLMTLRNFGMGKRSIEDRVQEEARCLVEELRKTKASPCDPTFILGCAPCNVICSIIFQKRFDYKDQQFLNLMEKLNENIRIVSTPWIQICNNFPTIIDYFPGTHNKLLKNLAFMESDILEKVKEHQESMDINNPRDFIDCFLIKMEKEKQNQQSEFTIENLVITAADLLGAGTETTSTTLRYALLLLLKHPEVTAKVQEEIERVIGRNRSPCMQDRGHMPYTDAVVHEVQRYIDLIPTSLPHAVTCDVKFRNYLIPKGTTILTSLTSVLHDNKEFPNPEMFDPRHFLDEGGNFKKSNYFMPFSAGKRICVGEGLARMELFLFLTFILQNFNLKSLIDPKDLDTTPVVNGFASVPPFYQLCFIPV\"\n",
    "#    WHERE gene_name = 'CYP2C19';\n",
    "#    ''')\n",
    "\n",
    "# Update edit_distance for each mutation\n",
    "#cursor.execute('SELECT mutation_id, ancestral_residue, variant_residue FROM Mutation')\n",
    "#mutations = cursor.fetchall()\n",
    "#for mutation_id, ancestral_residue, variant_residue in mutations:\n",
    "#    if variant_residue in amino_acid_mapping.get(ancestral_residue, []):\n",
    "#        edit_distance = 1\n",
    "#    else:\n",
    "#        edit_distance = 'NA'\n",
    "#    \n",
    "#    cursor.execute('''\n",
    "#        UPDATE Mutation\n",
    "#        SET edit_distance = ?\n",
    "#        WHERE mutation_id = ?\n",
    "#    ''', (edit_distance, mutation_id))\n",
    "\n",
    "\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f80c100f-addc-4e14-8ed5-d7c1876691f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Close the connection\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
